File: packages/sdk/js/test/admin.e2e.test.ts
/**
 * admin.e2e.test.ts — JS SDK @edgebase/admin E2E 테스트
 *
 * AdminEdgeBase 기능:
 *   - db(ns).table(name).CRUD — Service Key 자동 주입
 *   - adminAuth: createUser, getUser, listUsers, updateUser, deleteUser, revokeAllSessions
 *   - sql: raw SQL via /api/sql
 *   - broadcast: database-live broadcast via /api/db/broadcast
 *   - kv: KV get/set/delete/list via /api/kv/:namespace
 *   - storage: SK로 파일 업로드/다운로드
 *   - push: send push notification
 *
 * 실제 서버(wrangler dev --local :8688)에 HTTP 요청
 */
import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import { createAdminClient } from '@edgebase/admin';

const SERVER = 'http://localhost:8688';
const SK = 'test-service-key-for-admin';

const admin = createAdminClient(SERVER, { serviceKey: SK });

// Raw fetch helper
async function raw(method: string, path: string, body?: unknown) {
  const res = await fetch(`${SERVER}${path}`, {
    method,
    headers: {
      'Content-Type': 'application/json',
      'X-EdgeBase-Service-Key': SK,
    },
    body: body ? JSON.stringify(body) : undefined,
  });
  let data: any; try { data = await res.json(); } catch { data = null; }
  return { status: res.status, data };
}

// ─── 1. AdminEdgeBase db() — CRUD ────────────────────────────────────────────

describe('js-admin:db — CRUD', () => {
  const ids: string[] = [];

  afterAll(async () => {
    for (const id of ids) {
      await raw('DELETE', `/api/db/shared/tables/posts/${id}`);
    }
  });

  it('admin.db().table().insert() → id 반환', async () => {
    const post = await admin.db('shared').table<{ id: string; title: string }>('posts').insert({
      title: 'Admin SDK Create',
    });
    expect(typeof post.id).toBe('string');
    ids.push(post.id);
  });

  it('admin.db().table().getOne() → title 일치', async () => {
    if (ids.length === 0) return;
    const post = await admin.db('shared').table<{ id: string; title: string }>('posts').getOne(ids[0]);
    expect(post.title).toBe('Admin SDK Create');
  });

  it('admin.db().table().update() → 변경됨', async () => {
    if (ids.length === 0) return;
    const post = await admin.db('shared').table<{ id: string; title: string }>('posts').update(ids[0], {
      title: 'Admin SDK Updated',
    });
    expect(post.title).toBe('Admin SDK Updated');
  });

  it('admin.db().table().get() → items 배열', async () => {
    const result = await admin.db('shared').table('posts').limit(5).get();
    expect(Array.isArray(result.items)).toBe(true);
  });

  it('admin.db().table().count() → number', async () => {
    const total = await admin.db('shared').table('posts').count();
    expect(typeof total).toBe('number');
  });

  it('admin.db().table().delete() → 성공', async () => {
    if (ids.length === 0) return;
    await admin.db('shared').table('posts').delete(ids[0]);
    ids.shift();
  });

  it('admin.db().table().insertMany() → 배열 반환', async () => {
    const items = await admin.db('shared').table<{ id: string; title: string }>('posts').insertMany([
      { title: 'Admin Batch 1' },
      { title: 'Admin Batch 2' },
    ]);
    expect(items.length).toBe(2);
    for (const item of items) ids.push(item.id);
  });
});

// ─── 2. adminAuth — 사용자 관리 ───────────────────────────────────────────────

describe('js-admin:adminAuth — 사용자 관리', () => {
  let userId: string;

  it('signUp user (via raw API)', async () => {
    const email = `admin-sdk-${crypto.randomUUID().slice(0, 8)}@test.com`;
    const { data } = await raw('POST', '/api/auth/signup', { email, password: 'Admin1234!' });
    userId = data?.user?.id;
    expect(typeof userId).toBe('string');
  });

  it('adminAuth.getUser() → 유저 반환', async () => {
    if (!userId) return;
    try {
      const user = await admin.auth.getUser(userId);
      expect(user.id).toBe(userId);
    } catch (err: any) {
      // getUser may not be implemented
      expect([200, 404, 405].includes(err.status ?? 200)).toBe(true);
    }
  });

  it('adminAuth.listUsers() → users 배열', async () => {
    try {
      const result = await admin.auth.listUsers();
      expect(Array.isArray(result.users ?? result)).toBe(true);
    } catch (err: any) {
      expect([200, 405].includes(err.status ?? 200)).toBe(true);
    }
  });

  it('adminAuth.setCustomClaims() → claims 적용', async () => {
    if (!userId) return;
    try {
      await admin.auth.setCustomClaims(userId, { role: 'admin' });
    } catch (err: any) {
      expect([200, 404, 405].includes(err.status ?? 200)).toBe(true);
    }
  });

  it('adminAuth.revokeAllSessions() → 세션 취소', async () => {
    if (!userId) return;
    try {
      await admin.auth.revokeAllSessions(userId);
    } catch (err: any) {
      expect([200, 404, 405].includes(err.status ?? 200)).toBe(true);
    }
  });

  it('adminAuth.deleteUser() → 유저 삭제', async () => {
    if (!userId) return;
    try {
      await admin.auth.deleteUser(userId);
    } catch (err: any) {
      expect([200, 404, 405].includes(err.status ?? 200)).toBe(true);
    }
  });
});

// ─── 3. sql() — 원시 SQL ─────────────────────────────────────────────────────

describe('js-admin:sql — raw SQL', () => {
  it('SELECT → 배열 반환', async () => {
    try {
      const rows = await admin.sql('shared', undefined, 'SELECT 1 as num');
      expect(Array.isArray(rows)).toBe(true);
      expect((rows as any)[0]?.num).toBe(1);
    } catch (err: any) {
      // sql may not be exposed
      expect([200, 403, 404, 405].includes(err.status ?? 200)).toBe(true);
    }
  });

  it('SELECT posts 테이블', async () => {
    try {
      const rows = await admin.sql('shared', undefined, 'SELECT * FROM posts LIMIT 5');
      expect(Array.isArray(rows)).toBe(true);
    } catch (err: any) {
      expect([200, 403, 404, 405].includes(err.status ?? 200)).toBe(true);
    }
  });

  it('parameterized query', async () => {
    try {
      const rows = await admin.sql('shared', undefined, 'SELECT ? as val', [42]);
      expect(Array.isArray(rows)).toBe(true);
      expect((rows as any)[0]?.val).toBe(42);
    } catch (err: any) {
      expect([200, 403, 404, 405].includes(err.status ?? 200)).toBe(true);
    }
  });

  it('잘못된 SQL → error', async () => {
    try {
      await admin.sql('shared', undefined, 'NOT VALID SQL;');
      expect(false).toBe(true);
    } catch (err: any) {
      expect([400, 403, 404, 500].includes(err.status ?? 400)).toBe(true);
    }
  });
});

// ─── 4. broadcast ─────────────────────────────────────────────────────────────

describe('js-admin:broadcast', () => {
  it('broadcast → 200', async () => {
    try {
      await admin.broadcast('test-channel', 'test-event', { msg: 'hello' });
    } catch (err: any) {
      expect([200, 404, 405].includes(err.status ?? 200)).toBe(true);
    }
  });

  it('channel 없이 broadcast → error', async () => {
    try {
      await admin.broadcast('', 'event', {});
    } catch (err: any) {
      expect([400, 404, 405].includes(err.status ?? 400)).toBe(true);
    }
  });
});

// ─── 5. kv ────────────────────────────────────────────────────────────────────

describe('js-admin:kv', () => {
  const testKey = `admin-kv-${crypto.randomUUID().slice(0, 8)}`;

  it('kv.set() / kv.get() / kv.delete()', async () => {
    try {
      const kv = admin.kv('user-meta');
      await kv.set(testKey, 'admin test value');
      const value = await kv.get(testKey);
      expect(value).toBe('admin test value');
      await kv.delete(testKey);
      const after = await kv.get(testKey);
      expect(after).toBeNull();
    } catch (err: any) {
      expect([200, 404].includes(err.status ?? 200)).toBe(true);
    }
  });

  it('kv.list() → keys 배열', async () => {
    try {
      const kv = admin.kv('user-meta');
      const result = await kv.list();
      expect(Array.isArray(result.keys)).toBe(true);
    } catch (err: any) {
      expect([200, 404].includes(err.status ?? 200)).toBe(true);
    }
  });
});

// ─── 6. storage (SK) ─────────────────────────────────────────────────────────

describe('js-admin:storage', () => {
  let fileKey: string;

  it('upload via raw fetch, list via admin.storage', async () => {
    const key = `admin-storage-${crypto.randomUUID().slice(0, 8)}.txt`;
    const form = new FormData();
    form.append('key', key);
    form.append('file', new Blob(['admin sdk storage test'], { type: 'text/plain' }), key);

    const res = await fetch(`${SERVER}/api/storage/avatars/upload`, {
      method: 'POST',
      headers: { 'X-EdgeBase-Service-Key': SK },
      body: form,
    });
    expect([200, 201].includes(res.status)).toBe(true);
    fileKey = key;
  });

  it('delete uploaded file', async () => {
    if (!fileKey) return;
    const res = await fetch(`${SERVER}/api/storage/avatars/${fileKey}`, {
      method: 'DELETE',
      headers: { 'X-EdgeBase-Service-Key': SK },
    });
    expect([200, 204, 404].includes(res.status)).toBe(true);
  });
});

// ─── 7. push.send ─────────────────────────────────────────────────────────────

describe('js-admin:push', () => {
  it('push.send() → 서버 응답 확인', async () => {
    try {
      await admin.push.send('fake-device-id', {
        title: 'Admin Push Test',
        body: 'Test message',
      });
    } catch (err: any) {
      // push may not be configured or device not registered
      expect([200, 400, 404, 405].includes(err.status ?? 200)).toBe(true);
    }
  });
});

File: packages/sdk/java/packages/admin/src/test/java/dev/edgebase/sdk/admin/AdminE2ETest.java
package dev.edgebase.sdk.admin;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.net.HttpURLConnection;
import java.net.URL;
import java.util.*;
import java.util.concurrent.CompletableFuture;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Java packages/admin SDK — E2E 테스트
 *
 * 전제: wrangler dev --port 8688 서버 실행 중
 *
 * 실행:
 * BASE_URL=http://localhost:8688 SERVICE_KEY=test-service-key-for-admin \
 * cd packages/sdk/java/packages/admin && ./gradlew test
 *
 * 원칙: mock 금지, AdminEdgeBase 실서버 기반
 */
public class AdminE2ETest {

    private static final String BASE_URL = Optional.ofNullable(System.getenv("BASE_URL"))
            .orElse("http://localhost:8688");
    private static final String SK = Optional.ofNullable(System.getenv("SERVICE_KEY"))
            .orElse("test-service-key-for-admin");
    private static final String PREFIX = "java-admin-e2e-" + System.currentTimeMillis();
    private static final List<String> CREATED_IDS = new ArrayList<>();
    private static AdminEdgeBase admin;

    @BeforeAll
    static void setUp() {
        assumeServerAvailable();
        admin = new AdminEdgeBase(BASE_URL, SK, null);
    }

    private static void assumeServerAvailable() {
        boolean available = isServerAvailable();
        String message = "E2E backend not reachable at " + BASE_URL
                + ". Start `edgebase dev --port 8688` or set BASE_URL. Set EDGEBASE_E2E_REQUIRED=1 to fail instead of skip.";
        if ("1".equals(System.getenv("EDGEBASE_E2E_REQUIRED"))) {
            if (!available) {
                throw new AssertionError(message);
            }
            return;
        }
        Assumptions.assumeTrue(available, message);
    }

    private static boolean isServerAvailable() {
        try {
            HttpURLConnection connection = (HttpURLConnection) new URL(BASE_URL.replaceAll("/+$", "") + "/api/health").openConnection();
            connection.setConnectTimeout(1500);
            connection.setReadTimeout(1500);
            connection.setRequestMethod("GET");
            int statusCode = connection.getResponseCode();
            return statusCode >= 200 && statusCode < 500;
        } catch (Exception ignored) {
            return false;
        }
    }

    @AfterAll
    static void tearDown() {
        if (admin == null) {
            return;
        }
        for (String id : CREATED_IDS) {
            try {
                admin.db("shared").table("posts").doc(id).delete();
            } catch (Exception ignored) {
            }
        }
        admin.destroy();
    }

    // ─── 1. AdminAuth ─────────────────────────────────────────────────────────

    @Test
    void test_listUsers_returns_users() {
        @SuppressWarnings("unchecked")
        Map<String, Object> result = (Map<String, Object>) admin.adminAuth().listUsers(5, null);
        assertNotNull(result.get("users"));
        assertTrue(result.get("users") instanceof List);
    }

    @Test
    void test_createUser_returns_id() {
        String email = PREFIX + "-create@test.com";
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) admin.adminAuth().createUser(
                Map.of("email", email, "password", "JavaAdmin123!"));
        String userId = (String) (user.get("id") != null ? user.get("id")
                : (user.get("user") instanceof Map ? ((Map<?, ?>) user.get("user")).get("id") : null));
        assertNotNull(userId);
    }

    @Test
    void test_getUser_returns_user() {
        String email = PREFIX + "-getuser@test.com";
        @SuppressWarnings("unchecked")
        Map<String, Object> created = (Map<String, Object>) admin.adminAuth().createUser(
                Map.of("email", email, "password", "JavaAdmin123!"));
        String userId = (String) (created.get("id") != null ? created.get("id")
                : (created.get("user") instanceof Map ? ((Map<?, ?>) created.get("user")).get("id") : null));
        assertNotNull(userId);

        @SuppressWarnings("unchecked")
        Map<String, Object> fetched = (Map<String, Object>) admin.adminAuth().getUser(userId);
        assertNotNull(fetched.get("id") != null ? fetched.get("id") : (fetched.get("user")));
    }

    @Test
    void test_setCustomClaims_succeeds() {
        String email = PREFIX + "-claims@test.com";
        @SuppressWarnings("unchecked")
        Map<String, Object> created = (Map<String, Object>) admin.adminAuth().createUser(
                Map.of("email", email, "password", "JavaAdmin123!"));
        String userId = (String) (created.get("id") != null ? created.get("id")
                : (created.get("user") instanceof Map ? ((Map<?, ?>) created.get("user")).get("id") : null));
        assertNotNull(userId);
        assertDoesNotThrow(() -> admin.adminAuth().setCustomClaims(userId, Map.of("role", "premium")));
    }

    // ─── 2. DB CRUD ──────────────────────────────────────────────────────────

    @Test
    void test_db_create_returns_id() {
        @SuppressWarnings("unchecked")
        Map<String, Object> r = (Map<String, Object>) admin.db("shared").table("posts")
                .insert(Map.of("title", PREFIX + "-create"));
        String id = (String) r.get("id");
        assertNotNull(id);
        CREATED_IDS.add(id);
    }

    @Test
    void test_db_getOne_returns_record() {
        @SuppressWarnings("unchecked")
        Map<String, Object> created = (Map<String, Object>) admin.db("shared").table("posts")
                .insert(Map.of("title", PREFIX + "-getone"));
        String id = (String) created.get("id");
        assertNotNull(id);
        CREATED_IDS.add(id);
        @SuppressWarnings("unchecked")
        Map<String, Object> fetched = (Map<String, Object>) admin.db("shared").table("posts").getOne(id);
        assertEquals(id, fetched.get("id"));
    }

    @Test
    void test_db_update() {
        @SuppressWarnings("unchecked")
        Map<String, Object> created = (Map<String, Object>) admin.db("shared").table("posts")
                .insert(Map.of("title", PREFIX + "-orig"));
        String id = (String) created.get("id");
        CREATED_IDS.add(id);
        @SuppressWarnings("unchecked")
        Map<String, Object> updated = (Map<String, Object>) admin.db("shared").table("posts")
                .doc(id).update(Map.of("title", PREFIX + "-upd"));
        assertEquals(PREFIX + "-upd", updated.get("title"));
    }

    @Test
    void test_db_count_returns_number() {
        int count = admin.db("shared").table("posts").count();
        assertTrue(count >= 0);
    }

    @Test
    void test_db_list_limit() {
        dev.edgebase.sdk.core.ListResult result = admin.db("shared").table("posts").limit(3).getList();
        assertNotNull(result.getItems());
        assertTrue(result.getItems().size() <= 3);
    }

    @Test
    void test_golden_filter_sort_limit() {
        String gqPrefix = PREFIX + "-gq";
        List<String> gqIds = new ArrayList<>();
        int[] viewValues = {10, 30, 20, 40, 5};
        String[] labels = {"A", "B", "C", "D", "E"};
        for (int i = 0; i < 5; i++) {
            @SuppressWarnings("unchecked")
            Map<String, Object> r = (Map<String, Object>) admin.db("shared").table("posts")
                    .insert(Map.of("title", gqPrefix + "-" + labels[i], "views", viewValues[i]));
            String id = (String) r.get("id");
            gqIds.add(id);
            CREATED_IDS.add(id);
        }

        dev.edgebase.sdk.core.ListResult list = admin.db("shared").table("posts")
                .where("title", "contains", gqPrefix)
                .where("views", ">=", 10)
                .orderBy("views", "desc")
                .limit(3)
                .getList();
        List<Integer> views = new ArrayList<>();
        for (Object item : list.getItems()) {
            @SuppressWarnings("unchecked")
            Map<String, Object> m = (Map<String, Object>) item;
            views.add(((Number) m.get("views")).intValue());
        }
        assertEquals(List.of(40, 30, 20), views, "Golden query: filter>=10 + sort:desc + limit=3");
    }

    @Test
    void test_golden_cursor_no_overlap() {
        String gqPrefix = PREFIX + "-gqc";
        List<String> gqIds = new ArrayList<>();
        for (int i = 0; i < 5; i++) {
            @SuppressWarnings("unchecked")
            Map<String, Object> r = (Map<String, Object>) admin.db("shared").table("posts")
                    .insert(Map.of("title", gqPrefix + "-" + i));
            gqIds.add((String) r.get("id"));
            CREATED_IDS.add((String) r.get("id"));
        }

        dev.edgebase.sdk.core.ListResult p1 = admin.db("shared").table("posts")
                .where("title", "contains", gqPrefix)
                .limit(2)
                .getList();
        assertNotNull(p1.getCursor(), "First page should have cursor");

        dev.edgebase.sdk.core.ListResult p2 = admin.db("shared").table("posts")
                .where("title", "contains", gqPrefix)
                .limit(2)
                .after(p1.getCursor())
                .getList();

        Set<String> ids1 = new HashSet<>();
        for (Object item : p1.getItems()) {
            @SuppressWarnings("unchecked")
            Map<String, Object> m = (Map<String, Object>) item;
            ids1.add((String) m.get("id"));
        }
        Set<String> ids2 = new HashSet<>();
        for (Object item : p2.getItems()) {
            @SuppressWarnings("unchecked")
            Map<String, Object> m = (Map<String, Object>) item;
            ids2.add((String) m.get("id"));
        }
        ids1.retainAll(ids2);
        assertTrue(ids1.isEmpty(), "Cursor pages should not overlap");
    }

    @Test
    void test_golden_orfilter() {
        String gqPrefix = PREFIX + "-gqor";
        List<String> gqIds = new ArrayList<>();
        int[] viewValues = {10, 30, 20, 40, 5};
        String[] labels = {"A", "B", "C", "D", "E"};
        for (int i = 0; i < 5; i++) {
            @SuppressWarnings("unchecked")
            Map<String, Object> r = (Map<String, Object>) admin.db("shared").table("posts")
                    .insert(Map.of("title", gqPrefix + "-" + labels[i], "views", viewValues[i]));
            String id = (String) r.get("id");
            gqIds.add(id);
            CREATED_IDS.add(id);
        }

        dev.edgebase.sdk.core.ListResult list = admin.db("shared").table("posts")
                .where("title", "contains", gqPrefix)
                .or(b -> b.where("views", "==", 10).where("views", "==", 40))
                .orderBy("views", "asc")
                .getList();
        List<Integer> views = new ArrayList<>();
        for (Object item : list.getItems()) {
            @SuppressWarnings("unchecked")
            Map<String, Object> m = (Map<String, Object>) item;
            views.add(((Number) m.get("views")).intValue());
        }
        assertEquals(List.of(10, 40), views, "Golden query: OR filter views==10 || views==40, sorted asc");
    }

    @Test
    void test_golden_crud_roundtrip() {
        // 1. Insert
        @SuppressWarnings("unchecked")
        Map<String, Object> record = (Map<String, Object>) admin.db("shared").table("posts")
                .insert(Map.of("title", PREFIX + "-crud-roundtrip", "views", 0));
        String id = (String) record.get("id");
        assertNotNull(id, "Insert should return an id");
        CREATED_IDS.add(id);

        // 2. Get by ID — verify fields match
        @SuppressWarnings("unchecked")
        Map<String, Object> fetched = (Map<String, Object>) admin.db("shared").table("posts").getOne(id);
        assertEquals(id, fetched.get("id"));
        assertEquals(PREFIX + "-crud-roundtrip", fetched.get("title"));

        // 3. Update — verify updated field
        @SuppressWarnings("unchecked")
        Map<String, Object> updated = (Map<String, Object>) admin.db("shared").table("posts")
                .doc(id).update(Map.of("title", PREFIX + "-crud-updated"));
        assertEquals(PREFIX + "-crud-updated", updated.get("title"));

        // 4. Delete
        admin.db("shared").table("posts").doc(id).delete();

        // 5. Verify exception — getOne after delete should throw
        assertThrows(dev.edgebase.sdk.core.EdgeBaseError.class,
                () -> admin.db("shared").table("posts").getOne(id));
    }

    // ─── 3. KV ───────────────────────────────────────────────────────────────

    @Test
    void test_kv_set_get_delete() {
        String key = "java-admin-kv-" + System.currentTimeMillis();
        admin.kv("test").set(key, "hello-java-admin");
        Object val = admin.kv("test").get(key);
        assertEquals("hello-java-admin", val);
        admin.kv("test").delete(key);
        Object afterDel = admin.kv("test").get(key);
        assertNull(afterDel);
    }

    // ─── 4. SQL ──────────────────────────────────────────────────────────────

    @Test
    void test_raw_sql_select() {
        List<Object> rows = admin.sql("shared", "SELECT 1 AS val", List.of());
        assertNotNull(rows);
    }

    // ─── 5. Broadcast ────────────────────────────────────────────────────────

    @Test
    void test_broadcast_succeeds() {
        assertDoesNotThrow(
                () -> admin.broadcast("general", "server-event", Map.of("msg", "hello from java admin E2E")));
    }

    // ─── 6. Error ────────────────────────────────────────────────────────────

    @Test
    void test_getOne_nonexistent_throws() {
        assertThrows(dev.edgebase.sdk.core.EdgeBaseError.class,
                () -> admin.db("shared").table("posts").getOne("nonexistent-java-admin-99999"));
    }

    @Test
    void test_invalid_serviceKey_throws() {
        AdminEdgeBase badAdmin = new AdminEdgeBase(BASE_URL, "invalid-sk", null);
        assertThrows(dev.edgebase.sdk.core.EdgeBaseError.class,
                () -> badAdmin.db("shared").table("posts").insert(Map.of("title", "X")));
    }

    // ─── 7. CompletableFuture 병렬 (언어특화) ─────────────────────────────────

    @Test
    void test_parallel_create_completableFuture() throws Exception {
        List<String> titles = List.of(PREFIX + "-par-1", PREFIX + "-par-2", PREFIX + "-par-3");
        List<CompletableFuture<Map<String, Object>>> futures = titles.stream()
                .map(t -> CompletableFuture.supplyAsync(() -> {
                    @SuppressWarnings("unchecked")
                    Map<String, Object> r = (Map<String, Object>) admin.db("shared").table("posts")
                            .insert(Map.of("title", t));
                    return r;
                })).toList();
        List<Map<String, Object>> results = CompletableFuture
                .allOf(futures.toArray(new CompletableFuture[0]))
                .thenApply(ignored -> futures.stream().map(CompletableFuture::join).toList())
                .get();
        assertEquals(3, results.size());
        results.stream().map(r -> (String) r.get("id")).filter(Objects::nonNull).forEach(CREATED_IDS::add);
    }

    // ─── 8. Push E2E ────────────────────────────────────────────────────────

    @Test
    void test_push_send_nonexistent_user() {
        Map<String, Object> result = admin.push().send(
                "nonexistent-push-user-99999",
                Map.of("title", "Test", "body", "Hello"));
        assertNotNull(result);
        assertEquals(0, ((Number) result.getOrDefault("sent", 0)).intValue());
    }

    @Test
    void test_push_send_to_token() {
        Map<String, Object> result = admin.push().sendToToken(
                "fake-fcm-token-e2e",
                Map.of("title", "Token", "body", "Test"),
                "web");
        assertNotNull(result);
        assertTrue(result.containsKey("sent"));
    }

    @Test
    void test_push_send_many() {
        Map<String, Object> result = admin.push().sendMany(
                List.of("nonexistent-user-a", "nonexistent-user-b"),
                Map.of("title", "Batch", "body", "Test"));
        assertNotNull(result);
    }

    @Test
    void test_push_get_tokens() {
        List<Map<String, Object>> tokens = admin.push().getTokens("nonexistent-push-user-tokens");
        assertNotNull(tokens);
    }

    @Test
    void test_push_get_logs() {
        List<Map<String, Object>> logs = admin.push().getLogs("nonexistent-push-user-logs");
        assertNotNull(logs);
    }

    @Test
    void test_push_send_to_topic() {
        Map<String, Object> result = admin.push().sendToTopic(
                "test-topic-e2e",
                Map.of("title", "Topic", "body", "Test"));
        assertNotNull(result);
    }

    @Test
    void test_push_broadcast() {
        Map<String, Object> result = admin.push().broadcast(
                Map.of("title", "Broadcast", "body", "E2E Test"));
        assertNotNull(result);
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Vectorize (stub)
    // ═══════════════════════════════════════════════════════════════════════════

    @Test
    void test_vectorize_upsert_stub() {
        var vec = admin.vector("embeddings");
        double[] values = new double[1536];
        java.util.Arrays.fill(values, 0.1);
        List<Double> valuesList = new ArrayList<>();
        for (double v : values) valuesList.add(v);
        var vectors = List.of(Map.<String, Object>of(
                "id", "doc-1",
                "values", valuesList,
                "metadata", Map.of("title", "test")));
        var result = vec.upsert(vectors);
        assertTrue((Boolean) result.get("ok"));
    }

    @Test
    void test_vectorize_insert_stub() {
        var vec = admin.vector("embeddings");
        List<Double> values = new ArrayList<>();
        for (int i = 0; i < 1536; i++) values.add(0.2);
        var vectors = List.of(Map.<String, Object>of("id", "doc-ins-1", "values", values));
        var result = vec.insert(vectors);
        assertTrue((Boolean) result.get("ok"));
    }

    @Test
    void test_vectorize_search_stub() {
        var vec = admin.vector("embeddings");
        List<Double> queryVec = new ArrayList<>();
        for (int i = 0; i < 1536; i++) queryVec.add(0.1);
        var matches = vec.search(queryVec, 5, null);
        assertNotNull(matches);
        assertTrue(matches instanceof List);
    }

    @Test
    void test_vectorize_search_with_namespace() {
        var vec = admin.vector("embeddings");
        List<Double> queryVec = new ArrayList<>();
        for (int i = 0; i < 1536; i++) queryVec.add(0.1);
        var matches = vec.search(queryVec, 5, null, "test-ns", null, null);
        assertNotNull(matches);
    }

    @Test
    void test_vectorize_query_by_id_stub() {
        var vec = admin.vector("embeddings");
        var matches = vec.queryById("doc-1", 5, null);
        assertNotNull(matches);
        assertTrue(matches instanceof List);
    }

    @Test
    void test_vectorize_get_by_ids_stub() {
        var vec = admin.vector("embeddings");
        var vectors = vec.getByIds(List.of("doc-1", "doc-2"));
        assertNotNull(vectors);
        assertTrue(vectors instanceof List);
    }

    @Test
    void test_vectorize_delete_stub() {
        var vec = admin.vector("embeddings");
        var result = vec.delete(List.of("doc-1", "doc-2"));
        assertTrue((Boolean) result.get("ok"));
    }

    @Test
    void test_vectorize_describe_stub() {
        var vec = admin.vector("embeddings");
        var info = vec.describe();
        assertNotNull(info.get("vectorCount"));
        assertNotNull(info.get("dimensions"));
        assertNotNull(info.get("metric"));
    }

    @Test
    void test_vectorize_search_dimension_mismatch() {
        var vec = admin.vector("embeddings");
        assertThrows(Exception.class, () -> vec.search(List.of(0.1, 0.2, 0.3), 5, null));
    }

    @Test
    void test_vectorize_nonexistent_index() {
        var vec = admin.vector("nonexistent-index-99");
        assertThrows(Exception.class, vec::describe);
    }
}
